onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc

```diff
@@ -0,0 +1,508 @@
+
+        std::vector<int64_t> permutation;
+
+        for (int64_t i = 0; i < input_dims-2; ++i)
+          permutation.push_back(i);
+
+        permutation.push_back(input_dims-1);
+        permutation.push_back(input_dims-2);
+
+        emscripten::val options = emscripten::val::object();
+        options.set("permutation", emscripten::val::array(permutation));
+        b = model_builder.GetBuilder().call<emscripten::val>("transpose", b, options);
+      }
+
+      output = model_builder.GetBuilder().call<emscripten::val>("matmul", a, b);
+    }
+    break;
+
+  case RecognizedOperatorType::MatMulNhcw:
+  case RecognizedOperatorType::MatMulNhcwTransposeA:
+  case RecognizedOperatorType::MatMulNhcwTransposeB:
+    {
+      const size_t a_idx = 0, b_idx = 1;
+      emscripten::val a = model_builder.GetOperand(node.InputDefs()[a_idx]->Name());
+      emscripten::val b = model_builder.GetOperand(node.InputDefs()[b_idx]->Name());
+
+      emscripten::val options = emscripten::val::object();
+      std::vector<int64_t> permutation = {0,2,1,3};
+      std::vector<int64_t> permutation_a = {0,2,1,3};
+      std::vector<int64_t> permutation_b = {0,2,1,3};
+      if (recognized_operator_type == RecognizedOperatorType::MatMulNhcwTransposeA)
+      {
+        permutation_a = {0,2,3,1};
+      }
+      else if (recognized_operator_type == RecognizedOperatorType::MatMulNhcwTransposeB)
+      {
+        permutation_b = {0,2,3,1};
+      }
+
+      options.set("permutation", emscripten::val::array(permutation_a));
+      a = model_builder.GetBuilder().call<emscripten::val>("transpose", a, options);
+      options.set("permutation", emscripten::val::array(permutation_b));
+      b = model_builder.GetBuilder().call<emscripten::val>("transpose", b, options);
+      output = model_builder.GetBuilder().call<emscripten::val>("matmul", a, b);
+
+      options.set("permutation", emscripten::val::array(permutation));
+      output = model_builder.GetBuilder().call<emscripten::val>("transpose", output, options);
+    }
+    break;
+
+  case RecognizedOperatorType::ReduceSum:
+    {
+      auto kept_axes = m_components.back().GetLabels(m_label_indices);
+      assert(kept_axes.size() <= 1);
+      std::vector<uint32_t> reduced_axes;
+      // uint32_t kept_axes_mask = (kept_axes.size() << 0) - 1;
+      uint32_t kept_axes_mask = 0;
+      for (auto axis : kept_axes)
+      {
+          kept_axes_mask |= (1 << axis);
+      }
+      std::vector<int64_t> output_shape;
+      const auto& output_defs = node.OutputDefs();
+      ORT_RETURN_IF_NOT(GetShape(*output_defs[0], output_shape, logger), "Cannot get shape");
+      for (uint32_t axis = 0, axis_count = static_cast<uint32_t>(output_shape.size()); axis < axis_count; ++axis)
+      {
+          if (~kept_axes_mask & (1<<axis))
+          {
+              reduced_axes.push_back(axis);
+          }
+      }
+
+      emscripten::val input = model_builder.GetOperand(node.InputDefs()[0]->Name());
```

Comment:
Maybe we could centralize the first input outside the `switch` to simplified the code.

Reply:
`PairwiseOperandProcess` needs to get two inputs in function. Centralizing the first input outside the `switch` may break its "integrity". I prefer to decentralize them.